status/discover.go

```diff
@@ -91,66 +86,120 @@ func getPathURL(endpoint string) (path string) {
-			service = name
-			url = discoverServiceURL(tracingConfig.Namespace, name)
-			if url != "" {
-				break
-			}
-		}
+func discoverTracingService() (service string) {
+	client, err := getClient()
+	//  Return if there is a problem with the client
+	if err != nil {
+		log.Debugf("[TRACING] Service discovery failed: %v", err)
+		return
 	}
-
-	// The user has set the route or we found one in tracing or jaeger-query
-	if url != "" {
-		// Calculate if Path
-		path, err := checkIfQueryBasePath(tracingConfig.Namespace, service)
-		if err != nil {
-			log.Debugf("Error checking the query base path")
+	// Check for each service in list
+	for _, name := range tracingLookupRoutes {
+		service = name
+		// Try to discover the service
+		serv, err := client.GetService(config.Get().IstioNamespace, service)
+		// If there is no error and the service is not nil that means that we found tracing
+		if serv != nil && err == nil {
+			log.Debugf("[TRACING] Service in: %s", service)
+			break
+		} else {
+			// No service found set to empty for the last iteration
+			service = ""
 		}
-		// The user didn't set the URL, so we need to set
-		if tracingConfig.URL == "" {
-			tracingConfig.URL = url + path // Overwrite URL if the user didn't set
-		}
-
-		// We store the path
-		tracingConfig.Path = path
-
-		// Set the service
-		tracingConfig.Service = service
-		appstate.JaegerEnabled = true
 	}
+	return
+}
 
-	// Save config
-	appstate.JaegerConfig = tracingConfig
+func discoverTracingPath() (path string) {
+	tracingConfig := config.Get().ExternalServices.Tracing
+	// We had the service so we can check the Path
+	path, err := checkIfQueryBasePath(tracingConfig.Namespace, tracingConfig.Service)
+	if err != nil {
+		log.Debugf("[TRACING] Error checking the query base path")
+	}
+	return
+}
 
-	return tracingConfig.URL
+func discoverURLTracingService() (url string) {
+	// Try to discover the URL . Openshift Client
+	url, err := discoverServiceURL(appstate.JaegerConfig.Namespace, appstate.JaegerConfig.Service)
+	if err != nil {
+		log.Debugf("[TRACING] URL discovery failed: %v", err)
+		return
+	}
+	// Trim the string to format correctly the url with the path
+	url = strings.TrimSuffix(url, "/") + "/" + appstate.JaegerConfig.Path
+	appstate.JaegerEnabled = true
+	appstate.JaegerConfig.URL = url
+	return
 }
 
 func DiscoverJaeger() string {
+	// Kiali has all the configuration
 	if appstate.JaegerEnabled {
 		return appstate.JaegerConfig.URL
 	}
+
+	// Get the configuration
 	tracingConfig := config.Get().ExternalServices.Tracing
-	if tracingConfig.URL != "" && tracingConfig.Service != "" {
+
+	// There is not a service in the configuration we need discover the service
+	if tracingConfig.Service == "" {
+		log.Infof("[TRACING] Kiali Service Discovery")
+		service := discoverTracingService()
+		if service != "" {
+			tracingConfig.Service = service
+			log.Infof("[TRACING] Kiali found tracing service in %s", tracingConfig.Service)
```

Comment:
I wonder if that doesn't make too many logs as this kind of stuff is already logged from the `discoverServiceURL` function

Reply:
This should be online the first time, I hope that the user will set tracing to disable if he have not jaeger.

Reply:
btw i've seen that you added some logs for @skondkar ... but still i'm keeping my comment to not forget before merging